### DISCUSSION TITLE
JWT 토큰의 필드를 암호화하면 어떤 이점이 있고, 어떤 알고리즘이 사용될까?

### DISCUSSION THREAD

**Discussion by beatmeJY** [dailyge/dailyge-server#122] [Information]
## 🔗 연관 PR
- [[DLG-88] JWT 내부 필드를 암호화 한다.](https://github.com/dailyge/dailyge-server/pull/50)

<br/><br/>

## ❓ 작성 이유
먼저 [@devjun10님께서 공유해주신 내용](https://github.com/dailyge/dailyge-server/pull/9#discussion_r1644035316)인 JWT 토큰은 [여러 온라인 사이트](http://www.jwt.io/) 등에서 처럼 쉽게 해석할 수 있어 JWT에 사용자 정보가 들어 있다면 쉽게 노출될 가능성이 높습니다.

따라서 페이로드에 저장되고 있는 사용자 정보들은 모두 서버에서만 알 수 있도록 암호화하였는데, 이 부분에 대해서 팀원분들께서도 쉽게 이해할 수 있도록 작성해 보았습니다.

<br/><br/>

### 1. 중요한 부분만 암호화
- JWT는 Header, Payload Signature 이렇게 세 부분으로 나뉘어져 있습니다.
- 사용자의 정보가 담겨 있는 payload 부분만을 암호화하여 JWT가 너무 길어지지 않도록 하였습니다.
![image](https://github.com/user-attachments/assets/b4606c04-13d7-4c1d-a9b0-03b0442ea197)
- 만약 JWT 전체를 암호화 하게 될 경우, 위와 같이 모든 상황에서 JWT를 사용 시 복호화 과정이 필요하게 됩니다.

<br/><br/>

### 2. AES 대칭키 사용
- 암호화와 복호화 모두 서버에서 진행되기 때문에 대칭키를 사용하였습니다.

#### DES vs AES
- DES는 1970년대부터 설계되었고 56비트의 짧은 키 길이로 [1998년 56시간만에 암호를 해독하는 공격 하드웨어](https://www.techmonitor.ai/technology/eff_cracks_des_encryption_in_56_hours)를 만들정도로 보안에 취약합니다.
- DES는 라운드끼리의 의존성도 크지만 AES는 라운드가 독립적이라서 병렬처리의 이점도 더 활용하여 비교적 성능이 빠릅니다.
- 따라서 미국 정부가 채택한 이후로 널리 사용되고 있는 AES(고급 암호화 표준) 암호화 방식을 사용하였습니다.

#### AES 256 길이를 선택한 이유
- [해당 PR 리뷰 참조](https://github.com/dailyge/dailyge-server/pull/50#discussion_r1694635465)

<br/><br/>

### 3. CBC 구조로 암호화
암호화 구조에는 `EBC`, `CBC`, `CTR`, `OFB`, `CFB` 등의 다양한 구조들이 존재하는데,
그 중에서 저는 파일 및 데이터 암호화에서 가장 널리 사용되고 있는 `CBC`를 학습하여 암호화를 진행하였습니다.

![image](https://github.com/user-attachments/assets/9b7dd03e-8b3e-4b3b-9338-d95a936debfd)
- CBC는 Cipher Block Chaining Mode로 각 평문 블록이 `이전 암호문 블록`과 `XOR` 연산된 후에 암호화 되는 특징을 가지고 있습니다.
- 이는 이전 암호문이 다음 암호문 블록에 영향을 주기 때문에 더 안정적인 보안을 얻을 수 있게 됩니다.

<br/><br/>
#### ECB vs CBC
![image](https://github.com/user-attachments/assets/c6101005-9f7b-42e6-be3b-9b6cee78ea31)
- `ECB`의 암호화의 경우 블록간에 영향을 주지 않고 암호화를 진행하게 되기 때문에 항상 똑같은 `Key`를 가지고 암호화를 진행하게 됩니다.
- 따라서 `ECB`는 항상 같은 Key를 넣으면 같은 결과가 나오므로 위의 사진과 같이 BMP 파일을 암호화하여도 윤곽이 드러나게 됩니다.
- 하지만 `CBC`는 이전 암호문 블록이 다음 암호문 블록에 영향을 주기 때문에 가장 오른쪽 그림과 같이 형태를 알아볼 수 없습니다.

<br/><br/>

#### 랜덤 IV 사용(Initializing Vector)
- IV를 사용하는 이유는 `동일한 값`을 `동일한 키`로 암호화 하더라도 다른 암호화 된 결과 값을 얻을 수 있도록 사용합니다.
- 때문에 IV를 얼마나 랜덤한 값을 사용하는지에 따라 보안이 더 강화 될 수 있습니다.
- IV는 최초 블록에서 사용되지만 이전 블록의 결과가 이후의 블록 결과에 영향을 미치기 때문에 전혀 다른 결과값을 얻을 수 있게 됩니다.
- 저는 `java.util.Random`의 경우 시간 베이스로 난수를 생성하는 반면  `java.security.SecureRandom`는 OS의 임의 값을 베이스로 랜덤 값을 생성하기 때문에 난수 예측이 매우 힘든 `SecureRandom`로 IV를 생성하였습니다.
  - [[Oracle] Class Random](https://docs.oracle.com/javase/7/docs/api/java/util/Random.html)
  - [[Oracle] Class SecureRandom](https://docs.oracle.com/javase/7/docs/api/java/security/SecureRandom.html)

<br/><br/>

## ✏️ 추가 학습

`CBC` 이외에도 여러가지 암호화 구조가 존재하지만, `CTR`모드와 인증을 결합한 `GCM` 암호화 구조가 존재한다고 합니다.
`GCM`는 `CBC`의 병렬 처리가 불가능 한 단점을 보완하고, 암호화 과정에서 Hash 값을 추가하여 암호문의 무결성 검사도 가능하다고 합니다.
 - [How does AES GCM encryption work](https://crypto.stackexchange.com/questions/101106/how-does-aes-gcm-encryption-work)
 - [AES-GCM과 VAES 인스트럭션](https://wariua.github.io/facility/aes-gcm-and-vaes-instruction.html)
 - [Why would I ever use AES-256-CBC if AES-256-GCM is more secure?](https://security.stackexchange.com/questions/184305/why-would-i-ever-use-aes-256-cbc-if-aes-256-gcm-is-more-secure)

우선 암호화를 구현하는 것이 제가 처음이라 이번에는 비교적 학습자료가 많은 `CBC`로 구현하였는데,
추후 `CBC`와 `GCM` 두 가지를 직접 성능테스트하여 GCM을 사용하면 어느정도 빨라지는지 글을 작성해보겠습니다.🫡

<br/><br/>

## 📚 참조
 - [D2 안전한 패스워드 저장](https://d2.naver.com/helloworld/318732)
 - [안전한 암호화를 위한 AES 알고리즘에 대한 이해와 구현코드](https://brewagebear.github.io/aes-algorithm-and-chiper-mode/#step-2-aes256-%EC%95%8C%EA%B3%A0%EB%A6%AC%EC%A6%98%EC%9D%B4%EB%9E%80)

**Comment by devjun10** [comment]
좋네요 ㅎㅎ 잘 봤습니다! 재밌게 공부하셨네요. 👍
